### ISSUE TITLE
Bug: Middy, should `IdempotencyAlreadyInProgressError` return a 409?

### ISSUE THREAD

**Issue by andrewdavidcostello** [aws-powertools/powertools-lambda-typescript#3425]
### Expected Behavior

`IdempotencyAlreadyInProgressError` has a config option to be handled, like the the `responseHook` such as `onErrorResponseHook` or the error by default triggers a `409` as we make a positive assumption that as a transaction is in progress this request can be considered already processed at this point in time.

### Current Behavior

If `IdempotencyAlreadyInProgressError ` is triggered due to a parallel request by default an unhandled 500 error is thrown.

### Code snippet

```
export const handle = middy(handler)
  .use(makeHandlerIdempotent({ persistenceStore, config: idempotencyConfig }))
```

### Steps to Reproduce

1. Setup a lambda with middy `makeHandlerIdempotent` middleware
2. Create a parallel request to trigger a `IdempotencyAlreadyInProgressError`
3. See error in response `There is already an execution in progress with idempotency key: x` with status `500`

### Possible Solution

As above we make a positive assumption the in progress transaction is because a request is likely already processed and should return a 409 or functionality is provided to handle the error directly from the middleware.

I may be completely off in this report but it seems like 500 from something that is probable seems a bit of a harsh default?

### Powertools for AWS Lambda (TypeScript) version

latest

### AWS Lambda function runtime

22.x

### Packaging format used

npm

### Execution logs

```Shell

```

**Comment by dreamorosi** [comment]
Hi @andrewdavidcostello, thank you for opening the issue.

The Idempotency middleware is designed to do one of two things:
1. Return the response from your function, either by running it (new requests) or by returning the stored result (when idempotent)
2. Throw one of few errors depending on what the idempotency state is


When an error is thrown and it's not handled by your code, AWS Lambda interprets this as an unhandled exception or a server error.

By itself Lambda is just a compute environment, it doesn't inherently act as an API. Because of this, there's no HTTP code attached to these errors, you can check the codebase.

By reading both this issue, and the other one you opened, it seems that you're running an API with Lambda acting as the "server".

If this is the case, any unhandled error thrown by the function handler is interpreted by API Gateway as a 500 error, since from the caller perspective (API Gateway) the server (Lambda) just throws an error.

To customize what your API returns when certain errors occur, you'll have to either refactor your code to handle the error and return the status code you want, or work on the API side to transform the request.

**Comment by andrewdavidcostello** [comment]
Again thanks for the quick reply on this, much appreciated.

No worries, I guess the intent here is not to be overly opinionated which makes a lot of sense, we will update our error handler to handle these scenarios.

Just wanted to also be sure this was the intended behaviour.

**Comment by dreamorosi** [comment]
It's ok, it's a fair question, the problem is we don't have a way of influencing which HTTP error code is returned since we are not specifying one in the first place.

From the Idempotency middleware perspective, we purposefully don't want to have an opinion on how you handle the error since not all use cases will be APIs. But then again, even if we had an opinion, we still don't return a response with a status code in it, so we can't do much about this.

For this specific case, you can either write a custom Middy.js middleware similar to this:

```ts
import { IdempotencyAlreadyInProgressError } from '@aws-lambda-powertools/idempotency';

const idempotentErrorHandler = () => {
  const onError = async (request) => {
    if (request.error && request.error.name instanceof IdempotencyAlreadyInProgressError) {
      return {
        statusCode: 409,
        message: 'foo'
      }
    }
  };

  return {
    onError,
  };
};
```

or refactor the function code to make [only a portion of your code idempotent](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/idempotency/#indexts_1), rather than the entire request. This way you'd be able to wrap the idempotent part within a `try/catch` and return the error code you want.

**Comment by andrewdavidcostello** [comment]
Thanks, yeah the error handling that bubbles up to looks very similar to that just now so will stick with that.

If you don't mind a slight tangential query?

Is the Idempotency middleware meant to be able to prevent request spamming, as we observed for example sending 50 of the same request in quick succession that then ran against multiple lambda instances allowed multiple of the same requests to be created.

**Comment by dreamorosi** [comment]
> as we observed for example sending 50 of the same request in quick succession that then ran against multiple lambda instances allowed multiple of the same requests to be created.

No, this should not be the case - but if it has happened, we'd appreciate it if you could share a reproduction example that we can test. This way, if valid, we can try fixing the bug and prevent this from happening.
